include/util/multi_level_partition.hpp

```diff
@@ -0,0 +1,250 @@
+#ifndef OSRM_UTIL_MULTI_LEVEL_PARTITION_HPP
+#define OSRM_UTIL_MULTI_LEVEL_PARTITION_HPP
+
+#include "util/typedefs.hpp"
+#include "util/for_each_pair.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <numeric>
+#include <vector>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+namespace osrm
+{
+namespace util
+{
+namespace detail
+{
+// get the msb of an integer
+// return 0 for integers without msb
+template <typename T> std::size_t highestMSB(T value)
+{
+    static_assert(std::is_integral<T>::value, "Integer required.");
+    std::size_t msb = 0;
+    while (value > 0)
+    {
+        value >>= 1u;
+        msb++;
+    }
+    return msb;
+}
+}
+
+
+using LevelID = std::uint8_t;
+using CellID = std::uint32_t;
+
+static constexpr CellID INVALID_CELL_ID = std::numeric_limits<CellID>::max();
+
+// Mock interface, can be removed when we have an actual implementation
+class MultiLevelPartition
+{
+  public:
+    // Returns the cell id of `node` at `level`
+    virtual CellID GetCell(LevelID level, NodeID node) const = 0;
+
+    // Returns the highest level in which `first` and `second` are still in different cells
+    virtual LevelID GetHighestDifferentLevel(NodeID first, NodeID second) const = 0;
+
+    // Returns the level at which a `node` is relevant for a query from start to target
+    virtual LevelID GetQueryLevel(NodeID start, NodeID target, NodeID node) const = 0;
+
+    virtual std::size_t GetNumberOfLevels() const = 0;
+
+    virtual std::size_t GetNumberOfCells(LevelID level) const = 0;
+
+};
+
+class PackedMultiLevelPartition final : public MultiLevelPartition
```

Comment:
Would it make sense to check also a fixed length multi-level partitioning based on bisection IDs directly from the partitioner? Basically the question is about how and where translate bisection ids to `std::vector<std::vector<CellID>> partitions`